File: jcv-core/src/main/java/com/ekino/oss/jcv/core/validator/ValidatorIdInValueMatcher.java
/*
 * Copyright (c) 2019 ekino (https://www.ekino.com/)
 */
package com.ekino.oss.jcv.core.validator;

import com.ekino.oss.jcv.core.JsonContextMatcher;

import lombok.Getter;
import lombok.RequiredArgsConstructor;

import static java.util.Optional.*;

@RequiredArgsConstructor
class ValidatorIdInValueMatcher implements JsonContextMatcher {

    @Getter
    private final String validatorId;

    @Override
    public boolean matches(String prefix, Object expectedValue, Object actualValue) {

        return ofNullable(expectedValue)
            .filter(String.class::isInstance)
            .map(String.class::cast)
            .map(ValidatorTemplateManager::new)
            .map(ValidatorTemplateManager::extractId)
            .map(validatorId::equalsIgnoreCase)
            .orElse(false);
    }
}

File: jcv-core/src/main/kotlin/com/ekino/oss/jcv/core/validator/ValidatorIdInValueMatcher.kt
/*
 * Copyright (c) 2019 ekino (https://www.ekino.com/)
 */
package com.ekino.oss.jcv.core.validator

import com.ekino.oss.jcv.core.JsonContextMatcher

internal class ValidatorIdInValueMatcher(private val validatorId: String) : JsonContextMatcher {

    override fun matches(prefix: String, expectedValue: Any?, actualValue: Any?): Boolean {
        return expectedValue
            ?.takeIf { it is String }
            ?.let { ValidatorTemplateManager(it as String).extractId() }
            ?.let { validatorId.equals(it, ignoreCase = true) }
            ?: false
    }
}
